# Построение классификатора

Наш классификатор будет работать таким образом: по выборке строится дистанцированный граф, для него считается хроматическое число и размер максимального независимого множества, по ним и размеру выборки какой-то алгоритм классификации делает предсказание.

Пусть принятие нулевой гипотеры -- False, а первой -- True. Тогда ошибка первого рода это $FPR =\frac{FP}{FP + TN}$, 
а мощность это $Recall = \frac{TP}{TP + FN}$

Для начала сгенерируем данные для обучения классификатора

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from tqdm import tqdm
import pandas as pd
import sys
import os
project_root = os.path.abspath(os.path.join(os.getcwd(), '../..'))
sys.path.insert(0, project_root)

from src.utils.utils import monte_carlo_experiment_for_several_characteristics

Фиксируем параметры для построения графа

In [ ]:
D = 1.5                     # параметр d для dist графа
n_sampels = 20              # число итераций
lambda_h0 = 1               # lambda для H0
lambda_h1 = 1/np.sqrt(10)   # lambda для H1

In [ ]:
metrics_cromatic_number = []
metrics_size_maximal_independent_set = []
data_size = []
type_dist = []

for n in tqdm(range(10, 610, 1)):
    params_h0 = {
            'n': n,
            'x': D,
            'lambda': lambda_h0,
            'distribution': 'h0'
         }
    
    params_h1 = {
            'n': n,
            'x': D,
            'lambda': lambda_h1,
            'distribution': 'h1'
         }
    
    result_h0 = monte_carlo_experiment_for_several_characteristics(params_h0, n_sampels, False)
    result_h1 = monte_carlo_experiment_for_several_characteristics(params_h1, n_sampels, False)

    cromatic_number_h0 = result_h0['metrics_cromatic_number']
    size_maximal_independent_set_h0 = result_h0['metrics_size_maximal_independent_set']
    cromatic_number_h1 = result_h1['metrics_cromatic_number']
    size_maximal_independent_set_h1 = result_h1['metrics_size_maximal_independent_set']
   
    type_dist_h0 = [0 for i in range(n_sampels)]
    type_dist_h1 = [1 for i in range(n_sampels)]

    metrics_cromatic_number+=cromatic_number_h0
    metrics_cromatic_number+=cromatic_number_h1
    metrics_size_maximal_independent_set+=size_maximal_independent_set_h0
    metrics_size_maximal_independent_set+=size_maximal_independent_set_h1
    type_dist+=type_dist_h0
    type_dist+=type_dist_h1
    data_size+=[n for _ in range(2*n_sampels)]
    

data = {
    "cromatic_number": metrics_cromatic_number,
    "independent_set": metrics_size_maximal_independent_set,
    "data_size": data_size,
    "type_dist": type_dist
}


100%|██████████| 600/600 [30:25<00:00,  3.04s/it]


In [ ]:
df = pd.DataFrame(data) #наш датасет

Наши данные для обучения готовы.

## Посоздаем модельки

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import (accuracy_score, precision_score, 
                            recall_score, f1_score, 
                            confusion_matrix, classification_report)
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
import seaborn as sns


def fpr_score(y_true, y_pred, pos_label=1):
    """Вычисляет False Positive Rate (FPR) — долю ложноположительных срабатываний.    
    Параметры:
        y_true : array-like
            Истинные метки классов.
        y_pred : array-like
            Предсказанные метки классов.
        pos_label : int или str, default=1
            Метка положительного класса.
    
    Возвращает:
        float
            FPR (от 0 до 1). Если FP + TN = 0, возвращает 0.0.
    """
    tn, fp, _, _ = confusion_matrix(y_true, y_pred, labels=[1 - pos_label, pos_label]).ravel()
    denominator = fp + tn
    fpr = fp / denominator if denominator != 0 else 0.0
    return fpr


X = df.drop("type_dist", axis=1)
y = df['type_dist']

X_train, X_test, y_train, y_test = train_test_split(
    X, 
    y, 
    test_size=0.2,      # Размер тестовой выборки (20%)
    random_state=42,    # Для воспроизводимости
    stratify=y          # Сохраняем распределение классов
)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

### KNN

In [7]:
knn = KNeighborsClassifier(n_neighbors=5) 
knn.fit(X_train_scaled, y_train)
y_pred = knn.predict(X_test_scaled)

In [8]:
print("Оценка качества модели KNN:")
print("----------------------------------")
print(f"Recall: {recall_score(y_test, y_pred):.4f}")
print(f"FPR: {fpr_score(y_test, y_pred):.4f}")
print(f"Accuracy: {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision: {precision_score(y_test, y_pred):.4f}")
print(f"F1-score: {f1_score(y_test, y_pred):.4f}")

Оценка качества модели KNN:
----------------------------------
Recall: 0.9754
FPR: 0.0317
Accuracy: 0.9719
Precision: 0.9686
F1-score: 0.9720


Как мы видим, обычный KNN уже отлично справляется! Но попробуем другие модельки

### Логрегрессия

In [9]:
from sklearn.linear_model import LogisticRegression

logreg = LogisticRegression(
    penalty='l2',       
    C=1.0,             
    max_iter=1000,  
    random_state=42
)
logreg.fit(X_train_scaled, y_train)

y_pred = logreg.predict(X_test_scaled)


print("Оценка качества логистической регрессии:")
print("----------------------------------")
print(f"Recall: {recall_score(y_test, y_pred):.4f}")
print(f"FPR: {fpr_score(y_test, y_pred):.4f}")
print(f"Accuracy: {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision: {precision_score(y_test, y_pred):.4f}")
print(f"F1-score: {f1_score(y_test, y_pred):.4f}")

Оценка качества логистической регрессии:
----------------------------------
Recall: 0.9546
FPR: 0.0312
Accuracy: 0.9617
Precision: 0.9683
F1-score: 0.9614


Recall получился чуть ниже, чем у KNN

### Случайный лес

In [ ]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(
    n_estimators=100,       # Количество деревьев
    max_depth=None,         # Глубина деревьев (без ограничений)
    min_samples_split=2,    # Минимальное число образцов для разделения узла
    min_samples_leaf=1,     # Минимальное число образцов в листе
    max_features='sqrt',    # Количество признаков для поиска лучшего разделения
    random_state=42,
    n_jobs=-1,              # Использовать все ядра процессора
    class_weight='balanced' # Учет дисбаланса классов
)


with tqdm(total=100, desc="Training Random Forest") as pbar:
    rf.fit(X_train_scaled, y_train)
    pbar.update(100)


y_pred = rf.predict(X_test_scaled)


print("Оценка качества случайного леса:")
print("----------------------------------")
print(f"Recall: {recall_score(y_test, y_pred):.4f}")
print(f"FPR: {fpr_score(y_test, y_pred):.4f}")
print(f"Accuracy: {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision: {precision_score(y_test, y_pred):.4f}")
print(f"F1-score: {f1_score(y_test, y_pred):.4f}")



Training Random Forest: 100%|██████████| 100/100 [00:00<00:00, 206.87it/s]

Оценка качества случайного леса:
----------------------------------
Recall: 0.9700
FPR: 0.0342
Accuracy: 0.9679
Precision: 0.9660
F1-score: 0.9680


Метрики Recall и FPR получились хуже, чем у KNN

### Итого

| Метрика     | KNN  | LogReg      |  Random Forest  |
|-------------|----- |-------------|-----------------|
|FPR          |0.0317|   0.0312    |      0.0342     |
|Recall       |0.9754|   0.9546    |      0.9700     |
|Accuracy     |0.9719|   0.9617    |      0.9679     |

Лучше всего работает KNN. FPR чуть лучше у LogReg, но у нее сильно хуже Recall.

### Построение pipeline

In [ ]:
from src.utils.utils import build_dist_nx, calculate_chromatic_number, calculate_size_maximal_independent_set
import numpy as np
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.pipeline import Pipeline
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler


class GraphFeatureExtractor(BaseEstimator, TransformerMixin):
    """Экстрактор признаков для графов, совместимый с scikit-learn.
    
    Преобразует входные данные (координаты точек) в графы и вычисляет:
    - хроматическое число графа
    - размер максимального независимого множества
    - количество точек в исходных данных
    
    Параметры:
        d : float, default=1.5
            Пороговое расстояние для соединения вершин графа.
            Две точки соединяются ребром, если расстояние между ними <= d.
    
    Методы:
        transform(X, y=None):
            Преобразует массив координат в матрицу признаков графов.
        fit(X, y=None):
            Пустой метод для совместимости с scikit-learn API.
    """
    def __init__(self, d=1.5):
        self.d = d

    def transform(self, X, y=None):
        """Преобразует входные данные в признаки графов.
        
        Параметры:
            X : array-like
                Массив координат точек (формат: [[x1,y1], [x2,y2], ...])
            y : None
                Игнорируется, присутствует для совместимости
        
        Возвращает:
            np.ndarray
                Матрица признаков формы (n_samples, 3), где столбцы:
                1. Хроматическое число графа
                2. Размер максимального независимого множества
                3. Количество точек в исходных данных
        """
        features = []
        for data in X:
            G = build_dist_nx(data, self.d)
            chrom_num = calculate_chromatic_number(G)
            mis_size = calculate_size_maximal_independent_set(G)
            features.append([chrom_num, mis_size, len(data)])
        return np.array(features)

    def fit(self, X, y=None):
        """Пустой метод, необходимый для scikit-learn API."""
        return self


extractor = GraphFeatureExtractor(d=1.5)
extractor.fit(X,y)

# Pipeline: Graph features -> StandardScaler -> KNN
pipeline = Pipeline([
    ('graph_features', extractor),
    ('scaler', scaler),
    ('knn', knn)
])

Сгенерируем наборы данных для проверки нашего pipeline

In [ ]:
from src.utils.utils import DataGenerator

gen = DataGenerator()

test_data_25 = []
test_data_100 = []
test_data_500 = []
y_test = []

for _ in range(200):
    test_data_25.append(gen.generate_h0(25, lambda_h0))
    test_data_25.append(gen.generate_h1(25, lambda_h1))
    test_data_100.append(gen.generate_h0(100, lambda_h0))
    test_data_100.append(gen.generate_h1(100, lambda_h1))
    test_data_500.append(gen.generate_h0(500, lambda_h0))
    test_data_500.append(gen.generate_h1(500, lambda_h1))
    y_test.append(0)
    y_test.append(1)

In [ ]:
y_pred_25 = pipeline.predict(test_data_25)
y_pred_100 = pipeline.predict(test_data_100)
y_pred_500 = pipeline.predict(test_data_500)

Посмотрим на метрики нашего Pipeline

In [70]:
print("Оценка качества Pipeline для n=25:")
print("----------------------------------")
print(f"Recall: {recall_score(y_test, y_pred_25):.4f}")
print(f"FPR: {fpr_score(y_test, y_pred_25):.4f}")
print(f"Accuracy: {accuracy_score(y_test, y_pred_25):.4f}")
print(f"Precision: {precision_score(y_test, y_pred_25):.4f}")
print(f"F1-score: {f1_score(y_test, y_pred_25):.4f}")

print("Оценка качества Pipeline для n=100:")
print("----------------------------------")
print(f"Recall: {recall_score(y_test, y_pred_100):.4f}")
print(f"FPR: {fpr_score(y_test, y_pred_100):.4f}")
print(f"Accuracy: {accuracy_score(y_test, y_pred_100):.4f}")
print(f"Precision: {precision_score(y_test, y_pred_100):.4f}")
print(f"F1-score: {f1_score(y_test, y_pred_100):.4f}")

print("Оценка качества Pipeline для n=500:")
print("----------------------------------")
print(f"Recall: {recall_score(y_test, y_pred_500):.4f}")
print(f"FPR: {fpr_score(y_test, y_pred_500):.4f}")
print(f"Accuracy: {accuracy_score(y_test, y_pred_500):.4f}")
print(f"Precision: {precision_score(y_test, y_pred_500):.4f}")
print(f"F1-score: {f1_score(y_test, y_pred_500):.4f}")

Оценка качества Pipeline для n=25:
----------------------------------
Recall: 0.7050
FPR: 0.2550
Accuracy: 0.7250
Precision: 0.7344
F1-score: 0.7194
Оценка качества Pipeline для n=100:
----------------------------------
Recall: 0.9500
FPR: 0.0650
Accuracy: 0.9425
Precision: 0.9360
F1-score: 0.9429
Оценка качества Pipeline для n=500:
----------------------------------
Recall: 1.0000
FPR: 0.0100
Accuracy: 0.9950
Precision: 0.9901
F1-score: 0.9950


### Итого

Был сгенерирован набор обучающих и тестовых данных. На них обучены модели KNN, LogReg и RandomForest. Лучшие метрики на тестовых данных показала KNN. Далее мы построили pipeline, который по массиву данных любого размера предсказывает, каким распределением он сгенерирован.

Метрики:

| Метрика     |  25  |     100     |        500      |
|-------------|----- |-------------|-----------------|
|FPR          |0.2550|   0.0650    |      0.0100     |
|Recall       |0.7050|   0.9500    |      1.0000     |
|Accuracy     |0.7250|   0.9425    |      0.9950     |


Можно заметить, как с размером данных растут и показатели метрик. Это связано с тем, что при больших n данные разделяются куда лучше, чем при маленьких (графики есть в experiments_1).